ambry-router/src/main/java/com.github.ambry.router/DeleteOperation.java

```diff
@@ -162,48 +175,68 @@ void handleResponse(ResponseInfo responseInfo) {
+          updateOperationState(replica, routerErrorCode);
         } else {
           responseHandler.onRequestResponseError(replica, deleteResponse.getError());
-          processServerError(replica, deleteResponse.getError());
+          // The status of operation tracker will be updated within the processServerError method.
+          routerErrorCode = processServerError(replica, deleteResponse.getError());
         }
       } catch (IOException e) {
-        // @todo: Even this should really not happen. But we need a metric.
         logger.error("Unable to recover a deleteResponse from received stream.");
-        updateOperationState(replica, RouterErrorCode.UnexpectedInternalError);
+        routerErrorCode = RouterErrorCode.UnexpectedInternalError;
+        updateOperationState(replica, routerErrorCode);
       }
     }
     checkAndMaybeComplete();
+    if (routerErrorCode != null) {
+      dataNodeBasedMetrics.deleteRequestErrorCount.inc();
+      dataNodeBasedMetrics.countError(routerErrorCode);
+      dataNodeBasedMetrics.deleteRequestLatencyMs.update(time.milliseconds() - deleteRequestInfo.getStartTimeMs());
+    }
   }
 
   /**
    * A wrapper class that is used to check if a request has been expired.
    */
-  private class InflightRequestInfo {
-    final long submissionTime;
+  private class DeleteRequestInfo {
+    final long startTimeMs;
     final ReplicaId replica;
 
-    InflightRequestInfo(long submissionTime, ReplicaId replica) {
-      this.submissionTime = submissionTime;
+    DeleteRequestInfo(long submissionTime, ReplicaId replica) {
+      this.startTimeMs = submissionTime;
       this.replica = replica;
     }
+
+    long getStartTimeMs() {
+      return startTimeMs;
+    }
   }
 
   /**
    * Goes through the inflight request list of this {@code DeleteOperation} and remove those that
    * have been timed out.
    */
   private void cleanupExpiredInflightRequests() {
-    Iterator<Map.Entry<Integer, InflightRequestInfo>> itr = inflightRequestInfos.entrySet().iterator();
+    Iterator<Map.Entry<Integer, DeleteRequestInfo>> itr = deleteRequestInfos.entrySet().iterator();
     while (itr.hasNext()) {
-      InflightRequestInfo inflightRequestInfo = itr.next().getValue();
-      if (time.milliseconds() - inflightRequestInfo.submissionTime > routerConfig.routerRequestTimeoutMs) {
+      DeleteRequestInfo deleteRequestInfo = itr.next().getValue();
+      if (time.milliseconds() - deleteRequestInfo.startTimeMs > routerConfig.routerRequestTimeoutMs) {
         itr.remove();
-        updateOperationState(inflightRequestInfo.replica, RouterErrorCode.OperationTimedOut);
+        ReplicaId replica = deleteRequestInfo.replica;
+        DataNodeId dataNodeId = replica.getDataNodeId();
+        NonBlockingRouterMetrics.NodeLevelMetrics dataNodeBasedMetrics =
+            routerMetrics.getDataNodeBasedMetrics(dataNodeId);
+        dataNodeBasedMetrics.deleteRequestTimeoutCount.inc();
```

Comment:
I think having all these metrics is unnecessary. Could we try to be leaner? One could make a case for a lot of metrics, but we need to pick and choose. I don't really see a reason at this time that we would need metrics related to a specific type of request per datanode. Can we just keep request level metrics to datanodes?